### PR TITLE
Add stm32f4xx series serial driver 

### PR DESCRIPTION
**Pull request by xoviat**
This in an attempt to add USART send and receive functionality using the DMA controller on stm32f4xx series microcontrollers. I am unsure how successful this attempt will be, but I think most of the pieces are there. One difference between this attempt and the exisiting uarte is that I intend to use the waker_interrupt macro, which encapsulates the interrupt within the interface.

### CODE REVIEW

**Comment by xoviat** [comment]
A few points:
1. This uses the waker_interrupt macro to wake the future on an interrupt. Actually, though this is not my project, I would like to see a "universal" waker interrupt macro used thoroughout the project. The idea is that there would be one struct managing the interrupts, and futures could register to be woken on an interrupt. I may attempt to implement this universally if you want, because I think it's such an elegant design.

Note: in the even that a single interrupt is used for too many futures, a compromise could be to register a callback with the unified interrupt handler.

2. I haven't tested this yet, but I'm unsure where to put the examples for stm32. There should be some examples so that the usage is clear.

**Comment by xoviat** [comment]
@Dirbaio what about the examples?

**Comment by Dirbaio** [comment]
> I haven't tested this yet

Please do test it! I can't merge a PR with a code that has never been run ;)

> but I'm unsure where to put the examples for stm32. There should be some examples so that the usage is clear.

Yes, 100% agreed. Can you do an `embassy-stm32f4-examples` crate at the top level, similar to the `examples` one? I'll rename `examples` to `embassy-nrf-examples` for consistency after merging this.

File: embassy-stm32f4/src/lib.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/lib.rs:364]: `uarte` is the name of the UART peripheral in nrf52 chips. I don't think this is how stm32 calls it :) 

I'd follow the naming of existing stm32 crates, which I think is `serial`?

File: embassy-stm32f4/src/uarte.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/uarte.rs:75]: remove dead code in comments

File: embassy-stm32f4/src/uarte.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/uarte.rs:170]: minor nit: double negations are confusing
`!self.transfer.is_none()` -> `self.transfer.is_some()`

**Review by Dirbaio — COMMENTED** [review]
Commented a few items in the diff.

As discussed in #9, the drivers should use OwnedInterrupt, so user has to `take!()` the interrupt and pass it to `new`.

The driver should also be generic over Serial instances and also over DMA channels (is that possible on stm32? dma is different than on nrf!), I guess  you were already planning on doing that, but I'm pointing it out just in case :) 

Also, discussing with @timokroeger we've arrived at the conclusion that it's better to not use embedded-dma and accept the slight unsafety. This involves
- changing the futures to take `&[u8]` and `&mut [u8]`
- making `new` unsafe, with a doc comment saying "you may not drop futures"

**Comment by xoviat** [comment]
There are some other problems (I think solvable) with this approach, but the main problem with this approach is that the `SendFuture` and `ReceiveFuture` need to return `USART1` and `StreamX` resources to the main `Uarte` struct once they are finished. These resources are stored in the `Transfer` struct which provides a `free` method to free these resources. The problem is severalfold:

1. The transfer must be stored in the future because the buffer size varies; therefore we cannot fully know the type of the `Transfer` before we receive the buffer.
2. Storing the `Transfer` in the future using an `Option<Transfer>` allows us to return resources to the main `Uarte` struct theoretically, but renders us unable to call `is_done` in the main `poll()` method (because `self` is a `Pin`), and the only place to return resources is the drop method, which is later than I would like (I assume that resources should be returned in the poll method).

I apologize if the solution to this seems obvious; but I only recently started using rust. The other option (which now seems easier than  this route, which I took because it seemed easier at first), is to use a fixed size RingBuffer, which would allow us to fully determine the type of `Transfer`, which would allow us to store the `Transfer` in the `Uarte` struct; this seems to be something that I could implement.

One other option that I thought of is that it could be possible to modify the future in an interrupt handler, bypassing the `Pin`. I'm not quite sure whether that would be possible though.

------

After some poking around, I managed to put something together that passes the borrow checker. Maybe I could have some feedback on this, though. I'm still not sure if this works.

File: embassy-stm32f4/src/serial.rs
Comment (xoviat) [inline embassy-stm32f4/src/serial.rs:209]: Not completely sure what the implication of this is.

File: embassy-stm32f4/src/serial.rs
Comment (xoviat) [inline embassy-stm32f4/src/serial.rs:212]: I can't call this method, because the borrow checker complains. Need a solution to this.

File: embassy-stm32f4/src/serial.rs
Comment (xoviat) [inline embassy-stm32f4/src/serial.rs:224]: Another ugly set of lines to get around borrow checker complaints.

**Comment by xoviat** [comment]
> The driver should also be generic over Serial instances and also over DMA channels (is that possible on stm32? dma is different than on nrf!), I guess you were already planning on doing that, but I'm pointing it out just in case :)

To be clear, this is simple to implement for all once implemented on one, and I intend to do so.

**Comment by xoviat** [comment]
Okay, the send functionality is confirmed as working. The receive functionality is not implemented yet because there is a problem with the borrow checker with respect to returning the buffer.

**Comment by xoviat** [comment]
@Dirbaio Given the size of this, I'd like to restrict this PR to just the stm32405 for now and implement other mcus in a subsequent PR if others request it. I am just thinking about the other traits I need (mainly canbus and gpiote, it's called EXTI in stm32), and I'd rather work on those than other mcus at this point. Note: the API will be the same for all mcus; it only requires changes to embassy to add support. With respect to embedded-dma, I *have* to use it (I think) because the dma transfer API requires those traits to compile (I think, I may be mistaken). I do not want to duplicate code if I can avoid it, so I don't want to re-implement this with manually setting registers just to avoid embedded-hal. I plan to play around with the interrupts to resolve this issue:

> In case of transmitting we should await until transmission physically completed before returning Poll::Ready. Usually it happens after tx DMA buffer already is empty and when transfer complete (TC) event is triggered by transmitter, so this is last TC event for this transfer.
> This important particularly in case on simplex or half-duplex transfer when we need await until transmission actually complete to switch line to receive data.

However, that's probably all for now, aside from testing this on hardware, if that's acceptable for you.

**Comment by xoviat** [comment]
@Dirbaio let's put this in for now. I've confirmed the functionality on hardware, and while it could be better, I would like to have the folder structure in the repository so that I can make subsequent PRs.

File: embassy-stm32f4/src/interrupt.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/interrupt.rs:27]: The stm32f4 has 4 priority bits, so this should go from 0 to 15

File: embassy-stm32f4/src/interrupt.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/interrupt.rs:171]: what's the reason to leave these commented?

File: embassy-stm32f4/src/lib.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/lib.rs:311]: can you remove the dead commented code?

**Review by Dirbaio — CHANGES_REQUESTED** [review]

File: embassy-stm32f4/src/interrupt.rs
Comment (xoviat) [reply embassy-stm32f4/src/interrupt.rs:171]: They're on other stm32f4 series MCUs, but not stm32f405. I think they should be commented for now, until I implement the others.

File: embassy-stm32f4/src/interrupt.rs
Comment (Dirbaio) [inline embassy-stm32f4/src/interrupt.rs:48]: This needs updating with the extra levels too

**Pull request merged**
